File: src/mtc_pipeline/src/tool_exchange_stages.cpp
#include "mtc_pipeline/tool_exchange_stages.hpp"
#include <moveit/task_constructor/stages/move_to.h>
#include <cmath>

namespace {
constexpr double DOCK_SPACING_METERS = 0.1524;  // 6 inches between dock stations
}

ToolExchangeStages::ToolExchangeStages(const rclcpp::Node::SharedPtr& node)
    : BaseStages(node) {}

bool ToolExchangeStages::run(const mtc_pipeline::action::ToolExchangeAction::Goal& goal)
{
    nlohmann::json poses;
    try {
        poses = nlohmann::json::parse(goal.poses_json);
    } catch (const nlohmann::json::exception& e) {
        RCLCPP_ERROR(node()->get_logger(), "Failed to parse poses_json: %s", e.what());
        return false;
    }

    // Validate state transitions
    if (goal.operation == "load" && goal.current_attached_gripper != "none") {
        RCLCPP_ERROR(node()->get_logger(), "Cannot load %s: %s already attached",
            goal.gripper.c_str(), goal.current_attached_gripper.c_str());
        return false;
    }
    if (goal.operation == "dock" && goal.current_attached_gripper != goal.gripper) {
        RCLCPP_ERROR(node()->get_logger(), "Cannot dock %s: %s is attached",
            goal.gripper.c_str(), goal.current_attached_gripper.c_str());
        return false;
    }

    auto task = create_task_template(goal.operation == "load" ? "Load Tool" : "Dock Tool");
    auto sampling = make_pipeline_planner();
    auto cartesian = make_cartesian_planner();

    // Helper to add relative move stages
    auto add_move = [&](const std::string& label, const std::string& dir, double dist) {
        auto stage = create_relative_move_stage(label, dir, dist, cartesian);
        stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group", "ik_frame"});
        task.add(std::move(stage));
    };

    // Move to approach pose
    const auto& joint_pose = poses.at(goal.approach_pose);
    if (!joint_pose.is_array() || joint_pose.size() != 6) {
        RCLCPP_ERROR(node()->get_logger(), "'%s' must be array of 6 joint angles", goal.approach_pose.c_str());
        return false;
    }

    auto approach = std::make_unique<mtc::stages::MoveTo>("approach", sampling);
    approach->properties().configureInitFrom(mtc::Stage::PARENT, {"group", "ik_frame"});
    approach->setGroup(default_arm_group_name());
    approach->setGoal(joints_from_degrees(joint_pose.get<std::vector<double>>()));
    task.add(std::move(approach));

    // Lateral shift to align with dock (reference is dock 3)
    double offset_y = DOCK_SPACING_METERS * (3 - goal.dock_number);
    if (std::abs(offset_y) >= 1e-4) {
        add_move("shift to dock", offset_y >= 0 ? "right" : "left", std::abs(offset_y));
    }

    if (goal.operation == "load") {
        add_move("attach tool", "forward", 0.2);
        add_move("detach holder", "up", 0.15);
        add_move("retreat", "backward", 0.2);
    } else if (goal.operation == "dock") {
        add_move("align holder", "forward", 0.2);
        add_move("detach tool", "down", 0.15);
        add_move("retreat", "backward", 0.2);
    } else {
        RCLCPP_ERROR(node()->get_logger(), "Unknown operation: %s", goal.operation.c_str());
        return false;
    }

    return load_plan_execute(task);
}

File: src/mtc_py/mtc_py_lib/stages/tool_exchange_stages.py
"""ToolExchange stages - Python equivalent of tool_exchange_stages.hpp/cpp.

Tool exchange: load/dock grippers at magnetic holder stations.
Uses Cartesian moves for precise tool attachment/detachment.
"""

from moveit.task_constructor import stages
from mtc_py_lib.stages.base_stages import BaseStages, joints_from_degrees

# Tool exchange dock spacing (6 inches between dock stations)
DOCK_SPACING_METERS = 0.1524


class ToolExchangeStages(BaseStages):
    """Handles tool loading and docking operations at magnetic holders."""

    def run(self, goal) -> bool:
        """Execute ToolExchange action.

        Load sequence (attaching a tool):
        1. Move to approach pose
        2. Shift laterally to align with dock
        3. Move forward to attach tool
        4. Move up to detach from holder
        5. Retreat backward

        Dock sequence (storing a tool):
        1. Move to approach pose
        2. Shift laterally to align with dock
        3. Move forward to align with holder
        4. Move down to detach tool
        5. Retreat backward

        Args:
            goal: ToolExchangeAction.Goal with fields:
                - operation: "load" or "dock"
                - gripper: "hande", "epick", or "none" - gripper being loaded/docked
                - current_attached_gripper: What's currently attached
                - dock_number: Dock position (1-5)
                - approach_pose: Approach pose name
                - poses_json: Pose definitions from task

        Returns:
            True if successful, False otherwise
        """
        self.logger.info(
            f"Executing tool exchange: operation={goal.operation}, "
            f"gripper={goal.gripper}, dock={goal.dock_number}"
        )

        # Validate state transitions
        if goal.operation == "load" and goal.current_attached_gripper != "none":
            self.logger.error(
                f"Cannot load {goal.gripper}: "
                f"{goal.current_attached_gripper} already attached"
            )
            return False

        if goal.operation == "dock" and goal.current_attached_gripper != goal.gripper:
            self.logger.error(
                f"Cannot dock {goal.gripper}: "
                f"{goal.current_attached_gripper} is attached"
            )
            return False

        # Parse poses
        poses = self.parse_poses(goal.poses_json)
        if poses is None:
            return False

        task_name = "Load Tool" if goal.operation == "load" else "Dock Tool"
        task = self.create_task_template(task_name)
        sampling = self.make_pipeline_planner()
        cartesian = self.make_cartesian_planner()

        # 1. Move to approach pose
        joint_pose = self.get_joint_pose(poses, goal.approach_pose)
        if joint_pose is None:
            return False

        approach = stages.MoveTo("approach", sampling)
        approach.group = self.arm_group
        self._set_ik_frame(approach)
        approach.setGoal(joints_from_degrees(joint_pose))
        task.add(approach)

        # 2. Lateral shift to align with dock (reference is dock 3)
        offset_y = DOCK_SPACING_METERS * (3 - goal.dock_number)
        if abs(offset_y) >= 1e-4:
            direction = "right" if offset_y >= 0 else "left"
            shift = self.create_relative_move_stage(
                "shift to dock", direction, abs(offset_y), cartesian
            )
            task.add(shift)

        # 3-5. Perform load or dock sequence
        if goal.operation == "load":
            # Move forward to attach tool
            attach = self.create_relative_move_stage(
                "attach tool", "forward", 0.2, cartesian
            )
            task.add(attach)

            # Move up to detach from holder
            detach = self.create_relative_move_stage(
                "detach holder", "up", 0.15, cartesian
            )
            task.add(detach)

            # Retreat backward
            retreat = self.create_relative_move_stage(
                "retreat", "backward", 0.2, cartesian
            )
            task.add(retreat)

        elif goal.operation == "dock":
            # Move forward to align with holder
            align = self.create_relative_move_stage(
                "align holder", "forward", 0.2, cartesian
            )
            task.add(align)

            # Move down to detach tool
            detach = self.create_relative_move_stage(
                "detach tool", "down", 0.15, cartesian
            )
            task.add(detach)

            # Retreat backward
            retreat = self.create_relative_move_stage(
                "retreat", "backward", 0.2, cartesian
            )
            task.add(retreat)

        else:
            self.logger.error(f"Unknown operation: {goal.operation}")
            return False

        return self.load_plan_execute(task)
